packages/kmod-6.1-nvidia/kmod-6.1-nvidia.spec

```diff
@@ -250,7 +250,8 @@ popd
 %{_cross_libdir}/nvidia/tesla/libOpenCL.so.1.0.0
 %{_cross_libdir}/nvidia/tesla/libOpenCL.so.1
 %if "%{_cross_arch}" == "x86_64"
-%{_cross_libdir}/nvidia/tesla/libnvidia-compiler.so.%{tesla_ver}
+%{_cross_libdir}/nvidia/tesla/libnvidia-pkcs11.so.%{tesla_ver}
+%{_cross_libdir}/nvidia/tesla/libnvidia-pkcs11-openssl3.so.%{tesla_ver}
```

Comment:
Do these libraries have any runtime dependencies on `libssl` or similar?

Reply:
They depend on `libcrypto.so`:

```bash
NVIDIA-Linux-x86_64-535.54.03 ❯ readelf -d libnvidia-pkcs11-openssl3.so.535.54.03 | rg NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto.so.3]
NVIDIA-Linux-x86_64-535.54.03 ❯ readelf -d libnvidia-pkcs11.so.535.54.03 | rg NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libcrypto.so.1.1]
```

But, the linker should try to load `libcrypto.so` from the container's file system, right? At least `libnvidia-container` won't try to bind-mount `libcrypto` from the host's file system.

Reply:
As pointed out by Arnaldo those rely on openssl crypto libs. I am not quite sure what to make of this. Looking again through the archives dropped by nvidia there is no mention of these new libraries being added like they have for other changes in their documentation/change notes. The only hint is an extra licensing blurb in the README regarding using pkcs11 header files from the oasis open organization.

I will see if I can come up with some easy testing to validate what Arnaldo mentioned where the container file system would supply that dependency before merging.

Reply:
Digging around in those libraries, they seem to be almost empty "shells" for wrapping pkcs11 functionality. The only function they implement is pkcs11's `GetFunctionList` function to list available pkcs functions. Other than that it just redirects directly to `libcrypto` functions (one library each for `openssl 1.1`, and `openssl 3` implementation).

I can only speculate what these are for, but my bet would be on nvidia shipping libraries of the same name for other hardware that has offloading capabilities and does the crypto calculations on that hardware, while for tesla cards it merely resorts to the software `libcrypto` implementations. They would need the "shell"-library to have a uniform library interface that makes any example code work no matter what the hardware platform is.

Looking at `libnvidia-container` it works just as Arnaldo had pointed out. The container will have to provide necessary libraries outside of nvidia libraries from the container image, including `libcrypto`.